flink-ml-core/src/main/java/org/apache/flink/ml/common/broadcast/operator/AbstractBroadcastWrapperOperator.java

```diff
@@ -269,138 +200,162 @@ private OperatorMetricGroup createOperatorMetricGroup(
     }
 
     /**
-     * extracts common processing logic in subclasses' processing elements.
+     * Extracts common processing logic in subclasses' processing elements.
      *
-     * @param streamRecord the input record.
-     * @param inputIndex input id, starts from zero.
-     * @param elementConsumer the consumer function of StreamRecord, i.e.,
+     * @param streamRecord The input record.
+     * @param inputIndex Input id, starts from zero.
+     * @param elementConsumer The consumer function of StreamRecord, i.e.,
      *     operator.processElement(...).
-     * @param watermarkConsumer the consumer function of WaterMark, i.e.,
+     * @param watermarkConsumer The consumer function of WaterMark, i.e.,
      *     operator.processWatermark(...).
-     * @throws Exception possible exception.
+     * @param keyContextSetter The consumer function of setting key context, i.e.,
+     *     operator.setKeyContext(...).
+     * @throws Exception Possible exception.
      */
     @SuppressWarnings({"rawtypes", "unchecked"})
     protected void processElementX(
             StreamRecord streamRecord,
             int inputIndex,
             ThrowingConsumer<StreamRecord, Exception> elementConsumer,
-            ThrowingConsumer<Watermark, Exception> watermarkConsumer)
+            ThrowingConsumer<Watermark, Exception> watermarkConsumer,
+            ThrowingConsumer<StreamRecord, Exception> keyContextSetter)
             throws Exception {
-        if (!isBlocked[inputIndex]) {
-            if (areBroadcastVariablesReady()) {
-                if (hasPendingElements[inputIndex]) {
-                    processPendingElementsAndWatermarks(
-                            inputIndex, elementConsumer, watermarkConsumer);
-                    hasPendingElements[inputIndex] = false;
+        if (richContext.hasRichFunction) {
```

Comment:
It is not very readable to have 4 stacked `if` statements.

How about using something like this:

```
if (!richContext.hasRichFunction) {
    elementConsumer.accept(streamRecord);
} else if (richContext.isBlocked[inputIndex]) {
    while (!areBroadcastVariablesReady()) {
        richContext.mailboxExecutor.yield();
    }
    elementConsumer.accept(streamRecord);
} else if (!areBroadcastVariablesReady()) {
    richContext.dataCacheWriters[inputIndex].addRecord(
            CacheElement.newRecord(streamRecord.getValue()));
} else {
    if (richContext.hasPendingElements[inputIndex]) {
        processPendingElementsAndWatermarks(
                inputIndex, elementConsumer, watermarkConsumer, keyContextSetter);
        richContext.hasPendingElements[inputIndex] = false;
        keyContextSetter.accept(streamRecord);
    }
    elementConsumer.accept(streamRecord);
}
```

Reply:
The stacked `if` statements here represent the real execution logic here. For example in the code snippet you provided here, the `if` statement means that this operator does not need to access broadcast varaibles, while the three `else` statement are different conditions when this operator needs to access broadcast variables.

So the original code snippet seems more readable to me, what do you think?